File: csharp/src/Scon.Core/Scon.cs
namespace RealmLabs.Scon;

public static class Scon
{
    public static SconValue ParseString(string source) => new Resolver(new LoadOptions()).Eval(Parser.ParseDocument(source));
    public static SconValue ParseFile(string path, LoadOptions? options = null)
    {
        options ??= new LoadOptions();
        var file = Path.GetFullPath(path);
        var root = Path.GetFullPath(options.IncludeRoot ?? Path.GetDirectoryName(file)!);
        var source = File.ReadAllText(file);
        if (Encoding.UTF8.GetByteCount(source) > options.MaxFileSize) throw new SconException(ErrorCode.ResourceLimitExceeded, "maximum file size exceeded");
        var resolver = new Resolver(options with { IncludeRoot = root });
        resolver.Stack.Add(file); resolver.Seen.Add(file);
        return resolver.Eval(Parser.ParseDocument(source, file));
    }
    public static string FormatValue(SconValue value) => Formatter.FormatValue(value);
    public static SconValue GetPath(SconValue value, string path)
    {
        var current = value;
        foreach (var segment in path.Split('.')) { if (current is not SconObject obj) throw new SconException(ErrorCode.TypeMismatch, "path segment requires object"); current = obj.Get(segment) ?? throw new SconException(ErrorCode.MissingReference, "path is not defined"); }
        return current;
    }
}

File: java/scon-core/src/main/java/io/github/realmlabs/scon/Scon.java
package io.github.realmlabs.scon;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public final class Scon {
    private Scon() {}

    public static Analysis.ParsedSource parseSource(String source) {
        return Analysis.parseSource(source, null);
    }

    public static Analysis.DocumentAnalysis analyzeSource(String source) {
        return Analysis.analyzeSource(source, null);
    }

    public static SconValue parseString(String source) {
        return new Resolver(LoadOptions.defaults()).eval(Parser.parseDocument(source, null));
    }

    public static SconValue parseFile(Path path) {
        return parseFile(path, LoadOptions.defaults());
    }

    public static SconValue parseFile(Path path, LoadOptions options) {
        try {
            Path file = path.toAbsolutePath().normalize();
            Path root = options.includeRoot() == null ? file.getParent() : options.includeRoot().toAbsolutePath().normalize();
            String source = Files.readString(file);
            if (source.getBytes(java.nio.charset.StandardCharsets.UTF_8).length > options.maxFileSize()) {
                throw new SconException(ErrorCode.ResourceLimitExceeded, "maximum file size exceeded");
            }
            Resolver resolver = new Resolver(options.withIncludeRoot(root));
            resolver.stack.add(file);
            resolver.seen.add(file);
            return resolver.eval(Parser.parseDocument(source, file.toString()));
        } catch (IOException ex) {
            throw new SconException(ErrorCode.IncludeNotFound, "include file not found: " + ex.getMessage());
        }
    }

    public static String formatValue(SconValue value) {
        return Formatter.formatValue(value);
    }

    public static String formatSource(String source) {
        return SourceFormatter.formatSource(source);
    }

    public static SconValue getPath(SconValue value, String path) {
        SconValue current = value;
        for (String segment : path.split("\\.")) {
            if (!(current instanceof SconObject object)) {
                throw new SconException(ErrorCode.TypeMismatch, "path segment requires object");
            }
            current = object.get(segment);
            if (current == null) {
                throw new SconException(ErrorCode.MissingReference, "path is not defined");
            }
        }
        return current;
    }
}
